ext/drbbdup/drbbdup.c

```diff
@@ -708,29 +697,51 @@ drbbdup_encode_runtime_case(void *drcontext, drbbdup_per_thread *pt, void *tag,
     }
 
     /* Encoding is application-specific and therefore we need to user to define the
-     * encoding of the runtime case. We invoke a user-defined call-back.
-     */
-    ASSERT(opts.insert_encode, "The encode call-back cannot be NULL");
-    /* Note, we could tell the user not to reserve flags and scratch register since
-     * drbbdup is doing that already. However, for flexibility/backwards compatibility
-     * ease, this might not be the best approach.
+     * encoding of the runtime case. Therefore, we invoke a user-defined call-back.
+     *
+     * It could also be the encoding is done directly and changed on demand. Therefore,
+     * the call-back may be NULL.
      */
-    opts.insert_encode(drcontext, tag, bb, where, opts.user_data, pt->orig_analysis_data);
+    if (opts.insert_encode != NULL) {
 
-    /* Restore all unreserved registers used by the call-back. */
-    drreg_restore_all(drcontext, bb, where);
+        /* Note, we could tell the user not to reserve flags and scratch register since
+         * drbbdup is doing that already. However, for flexibility/backwards compatibility
+         * ease, this might not be the best approach.
+         */
+        opts.insert_encode(drcontext, tag, bb, where, opts.user_data,
+                           pt->orig_analysis_data);
 
-#ifdef X86_32
-    /* Load the encoding to the scratch register.
-     * The dispatcher could compare directly via mem, but this will
-     * destroy micro-fusing (mem and immed).
-     */
+        /* Restore all unreserved registers used by the call-back. */
+        drreg_restore_all(drcontext, bb, where);
+    }
+
+    /* Load the encoding to the scratch register. */
     opnd_t scratch_reg_opnd = opnd_create_reg(DRBBDUP_SCRATCH_REG);
-    opnd_t opnd = drbbdup_get_encoding_opnd();
-    instr_t *instr = INSTR_CREATE_mov_ld(drcontext, scratch_reg_opnd, opnd);
+    instr_t *instr =
+        INSTR_CREATE_mov_ld(drcontext, scratch_reg_opnd, opts.runtime_case_opnd);
```

Comment:
The issue for a globally changing case is memory visibility.  If I make my encoding address a single global, I would need a barrier here for aarchxx.  The alternative is to have the encoding be TLS and when the value changes go force all the threads to update their values -- although I'm not sure how to do that.  If I could run code as each thread I could have each go do an acquire load.  I can do that on UNIX by sending a signal.  I suppose I could do NtSetContextThread on Windows twice to run my code sequence.  I would want to add DR API support for those.  dr_suspend_all_other_threads() could do it I guess by setting the mcontext and then resuming -- but there's no per-thread control point to restore w/o having a wait point and another suspend-the-world.

I don't know how performance of a barrier at the top of a every block compares to a heavyweight interrupt-all-threads-when-change approach. Certainly the barrier is much much easier to implement (esp when #4215 is done).

How do you envision a typical use case changing the encoding?  Are all the use cases you're thinking of wanting local changes where only the current thread changes its own encoding independently of all others?

Should there then be an option for whether to use an acquire load or a regular load?  (This is why I was proposing a register interface before.)

Reply:
The interface could take a an option whether to use a lock. Another way is to support an additional call-back function that takes the memory operand and a register and loads the runtime encoding. The call-back is invoked after insert_encoding and drreg restoration are done.

Reply:
> How do you envision a typical use case changing the encoding? Are all the use cases you're thinking of wanting local changes where only the current thread changes its own encoding independently of all others?

Yep, all the ones I used are thread-local. I figured locking at the start of every basic block for global yields a performance hit that one would like the avoid.

Reply:
Does drbbdup read the encoding anywhere else?  I.e., could we remove the memory operand part of the interface and only have the register part.  So it would be back to the prior register proposal, except drbbdup picks the register instead of the user.

Reply:
The problem with that approach is the drreg restoration. The encoding is stored in memory so that drbbdup can restore *all* register values which were used by the user-defined insert_encoding call-back. This makes dispatch easier as the control landing of the appropriate basic block does not concern heavy register restoration.  If the user stores the encoding in a register and memory is not touched, then drbbdup will destroy the encoding.

Reply:
Still not really understanding the issue here.
It sounds like you would like to have two different callbacks and a provided memory address/operand X:

```
init: set mem/address/operand X
in bb:
  insert_encoding_to_memory_address_X_cb()
  restore all drreg lazily spilled regs by the above cb
  load_encoding_to_register_from_X_cb()
```

But wouldn't it be simpler to have:

```
<nothing in init>
in bb:
  load_encoding_to_register_from_anywhere_cb()
  restore all drreg lazily spilled regs by the above cb EXCEPT the target register
```

Just loop over the GPR's, skipping the one that holds the encoding.

Reply:
Do you mean create a new drreg API that un-reserves a register without ever restoring it back to its original value so that the encoding won't get clobbered? Basically, drbbdup maintains its own spill slot for the single target register in order to safely restore it after the dispatcher has directed control to the appropriate basic block case (drreg is not really designed for restoring registers when jumping across app code).

The situation is as follows:

```
init: set mem/address/operand X
in bb:
  insert_encoding_to_memory_address_X_cb() with the possibility of using ALL registers, maintained by drreg, to compute the encoding. 
  restore all drreg lazily spilled regs by the above cb < END DRREG USAGE WITH A CLEAN STATE>
  load_encoding_to_register_from_X_cb()  where the target register is not maintained by drreg.
```

Essentially I would like all registers to be maintained by drreg for the insert_encoding call-back so that the user can leverage the full drreg API, e.g. drreg_restore_app_value() including the target register.

Reply:
The drreg clean state can be between the cmp and jccc, with only the flags crossing the branch.

Reply:
What if you have multiple cases defined and therefore multiple compare and branch sequences?

Reply:
> > How do you envision a typical use case changing the encoding? Are all the use cases you're thinking of wanting local changes where only the current thread changes its own encoding independently of all others?
> 
> Yep, all the ones I used are thread-local. I figured locking at the start of every basic block for globals yields a performance hit that one would like the avoid.

Whether the cases are thread-local or global is part of the user-facing feature though: it's not just some internal implementation choice.  Do you have any suggestion for how to implement #3995 that does not involve either an acquire load (barrier to ensure stores from other threads are seen) on every single bb, or the scheme I described before of forcing all other threads to run such an acquire at the time of a mode change?

Reply:
> What if you have multiple cases defined and therefore multiple compare and branch sequences? Subsequent cmps won't have the runtime encoding available.

OK, though it feels like I didn't run into this kind of thing with the internal control flow in DrM: its general scheme is to reserve several regs once at the very top and don't unreserve until the very bottom where all control flow joins up for a single unreserve, with any extra regs needed completely local and not crossing branches (presumably with a barrier before a branch)(though DrM has its own register scheme and converting to drreg is not done and in an in-progress branch and I kept hitting feature I needed to add to drreg to make it work so maybe there is sthg that doesn't work still).  If this case-holding register is reserved across all the dispatch branches and throughout all the per-case code and only unreserved once at the bottom where all the cases converge back again does it work out for drbbdup.

Reply:
> > > How do you envision a typical use case changing the encoding? Are all the use cases you're thinking of wanting local changes where only the current thread changes its own encoding independently of all others?
> > 
> > 
> > Yep, all the ones I used are thread-local. I figured locking at the start of every basic block for globals yields a performance hit that one would like the avoid.
> 
> Whether the cases are thread-local or global is part of the user-facing feature though: it's not just some internal implementation choice. Do you have any suggestion for how to implement #3995 that does not involve either an acquire load (barrier to ensure stores from other threads are seen) on every single bb, or the scheme I described before of forcing all other threads to run such an acquire at the time of a mode change?

I would use locking for globals. I am not saying that drbbdup cannot be used for global cases, just that all the use-cases that I implemented were thread-local. However, I don't think that this is the issue we are facing.

In particular, the problem revolves around defining a suitable interface while dealing with the joint intricacies of drbbdup and other DR components such as drreg. I don't think there is an absolute correct answer, each design having its own  advantages and disadvantages.

- drbbdup maintains its own TLS slot and insert_encoding() stores the encoding to memory

(-) mem2mem copy, wasting a cache line.
(+)  insert_encoding() can leverage the full use of the drreg API.
(+) No complications regarding register restoration.
(+) Provides memory usage visibility, i.e., the call-back can easily perform a lock when loading the encoding.
(+) drbbdup book-keeping data per basic block can be stored in non reachable heap memory without the need to spill an additional register.

- insert_encoding() stores the encoding to a register

(-) insert_encoding() cannot leverage the full use of the drreg API or complications arise due to register restoration cross app-code.
(-) drbbdup book-keeping data per basic block has to be stored in non-reachable heap memory or another register needs to be spilled.
(+) Provides memory-usage-visibility
(+) No mem2mem copy

- drbbdup loads encoding from memory maintained by client.

(-) Problems with memory-usage-visibility since drbbdup internally performs the load.
(-) drbbdup book-keeping data per basic block has to be stored in non-reachable heap memory or another register needs to be spilled.
(+) No mem2mem copy
(+) No complications regarding register restoration.
(+)  insert_encoding() can leverage the full use of the drreg API.

Reply:
For my use case, performance is a priority, probably ruling out the first one.  The second would be fine for me since my encoding is very simple, but for a general solution maybe what you were proposing with the last one plus a load callback is the way to go.

Reply:
I added a new flag to the interface which informs drbbdup to perform the load locked/atomic.

Reply:
SGTM